README.md
# Deeper Pockets [![](http://cf.way2muchnoise.eu/full_deeper-pockets_downloads.svg)](https://minecraft.curseforge.com/projects/deeper-pockets)

Extremely simple miners backpacks.

Deeper Pockets is a fork of [DeepPockets](https://github.com/RebelKeithy/Deep-Pockets) by RebelKeithy and is allowed under the MIT license.

## Features

Deeper Pockets adds different backpacks that store items in a very loose way. Rather than giving you specific numbers, you are given a general idea.

![](https://cdn.discordapp.com/attachments/375860824457084931/422195498510843905/unknown.png)

In order to retrieve the items from your backpack, sneak right click an inventory to insert the items. They will be placed randomly throughout.

![](https://cdn.discordapp.com/attachments/375860824457084931/422195943962705950/unknown.png)

Currently there are only 2 tiers of Miner's Backpacks. The Miner's Backpack (which can store 5 types) and the Advanced Miner's Backpack (which can store 10).

| Prefix     | Amount          |
|------------|-----------------|
| No         | 0               |
| A few      | 1 to 4          |
| Several    | 5 to 15         |
| Piles of   | 16 to 31        |
| Lots of    | 32 to 63        |
| Hoards of  | 64 to 127       |
| Throngs of | 128 to 255      |
| Swarms of  | 256 to 511      |
| Zounds of  | 512 to 1023     |
| Legions of | 1024 and higher |

changelog.txt
------------------------------------------------------
Version 1.1.0
------------------------------------------------------

- Initial fork release

src/main/java/com/rebelkeithy/deeppockets/EventHandler.java
package com.rebelkeithy.deeppockets;

import com.rebelkeithy.deeppockets.item.ItemMiningPack;
import com.rebelkeithy.deeppockets.item.ItemMiningPack.StoredOre;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.init.SoundEvents;
import net.minecraft.item.ItemStack;
import net.minecraft.util.SoundCategory;
import net.minecraftforge.event.entity.player.EntityItemPickupEvent;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;

import java.util.Random;

@Mod.EventBusSubscriber(modid = DeepPockets.MODID)
public class EventHandler {

    @SubscribeEvent
    public static void pickupItemEvent(EntityItemPickupEvent event) {
        if (event.getEntityPlayer() != null) {
            String oreName = DeepPocketsConfig.isOre(event.getItem().getItem());
            if (oreName != null) {
                EntityPlayer player = event.getEntityPlayer();
                for (ItemStack stack : player.inventory.mainInventory) {
                    if (stack.getItem() instanceof ItemMiningPack) {
                        ItemMiningPack pack = (ItemMiningPack) stack.getItem();

                        if (pack.numStoredOres(stack) < pack.getType().getSlots() || pack.containsOre(stack, oreName)) {
                            StoredOre ore = pack.getOre(stack, oreName);
                            ore.amount += event.getItem().getItem().getCount();
                            ore.displayName = event.getItem().getItem().getDisplayName();
                            if (ore.registryName == null) {
                                ore.registryName = event.getItem().getItem().getItem().getRegistryName().toString();
                                ore.meta = event.getItem().getItem().getItemDamage();
                            }
                            pack.setOre(stack, oreName, ore);

                            Random rand = new Random();
                            float f = ((rand.nextFloat() - rand.nextFloat()) * 0.7F + 1.0F) * 2.0F;
                            player.world.playSound(null, player.posX, player.posY, player.posZ, SoundEvents.ENTITY_ITEM_PICKUP, SoundCategory.AMBIENT, 0.2f, f);

                            event.getItem().setDead();
                            event.setCanceled(true);
                            return;
                        }
                    }
                }
            }
        }
    }
}


src/main/java/com/rebelkeithy/deeppockets/DeepPocketsConfig.java
package com.rebelkeithy.deeppockets;

import com.google.common.collect.Lists;
import com.rebelkeithy.deeppockets.item.ItemMiningPack.StoredOre;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.common.config.Config;
import net.minecraftforge.common.config.ConfigManager;
import net.minecraftforge.fml.client.event.ConfigChangedEvent;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.registry.ForgeRegistries;
import net.minecraftforge.oredict.OreDictionary;

import java.util.Collections;
import java.util.List;

@Config(modid = DeepPockets.MODID)
@Mod.EventBusSubscriber(modid = DeepPockets.MODID)
public class DeepPocketsConfig {

    @Config.Comment({ "Items other than ore blocks" })
    public static String[] allowedItems = new String[] {
            "minecraft:coal",
            "minecraft:redstone",
            "minecraft:dye:4",
            "minecraft:diamond",
            "minecraft:emerald",
            "minecraft:quartz"
    };
    @Config.Ignore
    public static List<String> builtItems = Lists.newArrayList();

    @SubscribeEvent
    public static void onConfigChanged(ConfigChangedEvent.OnConfigChangedEvent event) {
        if (event.getModID().equals(DeepPockets.MODID)) {
            ConfigManager.sync(DeepPockets.MODID, Config.Type.INSTANCE);
            buildItemList();
        }
    }

    public static void buildItemList() {
        builtItems.clear();
        Collections.addAll(builtItems, allowedItems);
    }

    // Returns null if not an ore, otherwise returns the oredict name
    public static String isOre(ItemStack stack) {
        if (builtItems.contains(stack.getItem().getRegistryName().toString()))
            return stack.getItem().getRegistryName().toString();

        if (builtItems.contains(stack.getItem().getRegistryName().toString() + ":" + stack.getMetadata()))
            return stack.getItem().getRegistryName().toString() + ":" + stack.getMetadata();

        int[] ids = OreDictionary.getOreIDs(stack);
        if (ids.length > 0) {
            String name = OreDictionary.getOreName(ids[0]);
            return name.contains("ore") ? name : null;
        }

        return null;
    }

    public static ItemStack getStack(StoredOre ore, int amount) {
        Item item = ForgeRegistries.ITEMS.getValue(new ResourceLocation(ore.registryName));
        return new ItemStack(item, amount, ore.meta);
    }
}


src/main/java/com/rebelkeithy/deeppockets/item/RegistrarDeepPockets.java
package com.rebelkeithy.deeppockets.item;

import com.rebelkeithy.deeppockets.DeepPockets;
import com.rebelkeithy.deeppockets.DeepPocketsConfig;
import net.minecraft.client.renderer.ItemMeshDefinition;
import net.minecraft.client.renderer.block.model.ModelResourceLocation;
import net.minecraft.init.Items;
import net.minecraft.item.Item;
import net.minecraftforge.client.event.ModelRegistryEvent;
import net.minecraftforge.client.model.ModelLoader;
import net.minecraftforge.event.RegistryEvent;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.registry.GameRegistry;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

@Mod.EventBusSubscriber(modid = DeepPockets.MODID)
@GameRegistry.ObjectHolder(DeepPockets.MODID)
public class RegistrarDeepPockets {

    public static final Item MINING_PACK = Items.AIR;
    public static final Item MINING_PACK_ADVANCED = Items.AIR;

    @SubscribeEvent
    public static void registerItems(RegistryEvent.Register<Item> event) {
        DeepPocketsConfig.buildItemList();

        event.getRegistry().registerAll(
                new ItemMiningPack(PackTypes.BASIC).setRegistryName("mining_pack"),
                new ItemMiningPack(PackTypes.ADVANCED).setRegistryName("mining_pack_advanced")
        );
    }

    @SideOnly(Side.CLIENT)
    @SubscribeEvent
    public static void registerModels(ModelRegistryEvent event) {
        final ItemMeshDefinition packDefinition = stack -> {
            if (!(stack.getItem() instanceof ItemMiningPack))
                return new ModelResourceLocation("potato", "inventory"); // Shrug

            ItemMiningPack miningPack = (ItemMiningPack) stack.getItem();
            int stored = miningPack.getTotalItems(stack);
            int modelIndex = Math.min(1 + stored / 128, 4);
            return new ModelResourceLocation(stack.getItem().getRegistryName().toString() + "_" + modelIndex, "inventory");
        };

        for (int i = 1; i <= 4; i++) {
            ModelLoader.registerItemVariants(MINING_PACK, new ModelResourceLocation(MINING_PACK.getRegistryName().toString() + "_" + i, "inventory"));
            ModelLoader.registerItemVariants(MINING_PACK_ADVANCED, new ModelResourceLocation(MINING_PACK_ADVANCED.getRegistryName().toString() + "_" + i, "inventory"));
        }

        ModelLoader.setCustomMeshDefinition(MINING_PACK, packDefinition);
        ModelLoader.setCustomMeshDefinition(MINING_PACK_ADVANCED, packDefinition);
    }
}


src/main/java/com/rebelkeithy/deeppockets/item/ItemMiningPack.java
package com.rebelkeithy.deeppockets.item;

import com.google.common.collect.Lists;
import com.google.common.collect.Maps;
import com.rebelkeithy.deeppockets.DeepPocketsConfig;
import com.rebelkeithy.deeppockets.DeepPockets;
import net.minecraft.block.state.IBlockState;
import net.minecraft.client.resources.I18n;
import net.minecraft.client.util.ITooltipFlag;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.init.Blocks;
import net.minecraft.init.SoundEvents;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.EnumActionResult;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.SoundCategory;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.text.TextFormatting;
import net.minecraft.world.World;
import net.minecraftforge.items.CapabilityItemHandler;
import net.minecraftforge.items.IItemHandler;
import net.minecraftforge.items.IItemHandlerModifiable;
import net.minecraftforge.items.ItemHandlerHelper;
import net.minecraftforge.oredict.OreDictionary;

import javax.annotation.Nullable;
import java.util.*;

public class ItemMiningPack extends Item {

    public static final TreeMap<Integer, String> PREFIX = Maps.newTreeMap();

    static {
        PREFIX.put(0, "no");
        PREFIX.put(1, "few");
        PREFIX.put(5, "several");
        PREFIX.put(16, "piles");
        PREFIX.put(32, "lots");
        PREFIX.put(64, "hoards");
        PREFIX.put(128, "throngs");
        PREFIX.put(256, "swarms");
        PREFIX.put(512, "zounds");
        PREFIX.put(1024, "legions");
    }

    private final PackTypes type;

    public ItemMiningPack(PackTypes type) {
        this.type = type;

        setUnlocalizedName(DeepPockets.MODID + ".mining_pack_" + type.getName());
        setMaxStackSize(1);
        setCreativeTab(CreativeTabs.TOOLS);
    }

    public PackTypes getType() {
        return type;
    }

    @Override
    public void addInformation(ItemStack stack, @Nullable World worldIn, List<String> tooltip, ITooltipFlag flagIn) {
        NBTTagCompound tag = stack.getTagCompound();
        if (tag == null)
            return;

        NBTTagCompound ores = tag.getCompoundTag("ores");

        if (ores.getKeySet().size() > 0)
            tooltip.add(TextFormatting.GRAY + I18n.format("tooltip.deeppockets.slots_used", ores.getSize(), type.getSlots()));

        for (String ore : ores.getKeySet()) {
            int amount = ores.getCompoundTag(ore).getInteger("amount");
            String name = ores.getCompoundTag(ore).getString("name");
            String prefix = "prefix.deeppockets." + PREFIX.ceilingEntry(Math.min(amount, 1024)).getValue();

            String lock = "";
            if (ores.getCompoundTag(ore).getBoolean("lock"))
                lock = TextFormatting.DARK_GRAY + " *";

            tooltip.add(TextFormatting.GRAY + I18n.format(prefix, TextFormatting.DARK_AQUA + name.replace(" Ore", "") + lock)); // FIXME This Ore replacement will not work in other langs
        }
    }

    /**
     * Called when a Block is right-clicked with this Item
     */
    @Override
    public EnumActionResult onItemUse(EntityPlayer player, World worldIn, BlockPos pos, EnumHand hand, EnumFacing facing, float hitX, float hitY, float hitZ) {
        TileEntity te = worldIn.getTileEntity(pos);
        if (te != null && te.hasCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, facing)) {
            IItemHandler itemHandler = te.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, facing);
            if (itemHandler instanceof IItemHandlerModifiable) {

                ItemStack itemStack = player.getHeldItem(hand);

                if (worldIn.isRemote) {
                    if (!isEmpty(itemStack))
                        worldIn.playSound(player, pos, SoundEvents.BLOCK_GRAVEL_FALL, SoundCategory.PLAYERS, 0.5f, 0.6f);
                    return EnumActionResult.PASS;
                }
                return this.depositOresRandomly((IItemHandlerModifiable) itemHandler, itemStack);
            }
        }


        IBlockState block = worldIn.getBlockState(pos);
        ItemStack clickedBlockStack = block.getBlock().getPickBlock(block, null, worldIn, pos, player);
        if (block.getBlock() == Blocks.LIT_REDSTONE_ORE)
            clickedBlockStack = new ItemStack(Blocks.REDSTONE_ORE, 1);

        if (clickedBlockStack.isEmpty())
            return EnumActionResult.PASS;

        int[] oreIDs = OreDictionary.getOreIDs(clickedBlockStack);
        if (oreIDs.length > 0) {
            String oreName = OreDictionary.getOreName(oreIDs[0]);
            if (oreName.contains("ore")) {
                ItemStack itemStack = player.getHeldItemMainhand();
                NBTTagCompound tag = itemStack.getTagCompound();
                if (tag == null)
                    tag = new NBTTagCompound();

                NBTTagCompound ores = tag.getCompoundTag("ores");
                NBTTagCompound thisOre = ores.getCompoundTag(oreName);
                if (thisOre.getBoolean("lock")) {
                    thisOre.setBoolean("lock", false);
                    worldIn.playSound(player, pos, SoundEvents.ENTITY_EXPERIENCE_ORB_PICKUP, SoundCategory.PLAYERS, 0.3f, 0.5f);
                } else if (ores.getSize() < type.getSlots()) {
                    thisOre.setBoolean("lock", true);
                    thisOre.setString("name", clickedBlockStack.getDisplayName());
                    thisOre.setString("registry", clickedBlockStack.getItem().getRegistryName().toString());
                    thisOre.setInteger("meta", clickedBlockStack.getItemDamage());
                    if (!thisOre.hasKey("amount"))
                        thisOre.setInteger("amount", 0);
                    worldIn.playSound(player, pos, SoundEvents.ENTITY_EXPERIENCE_ORB_PICKUP, SoundCategory.PLAYERS, 0.3f, 0.8f);
                }


                ores.setTag(oreName, thisOre);
                if (thisOre.getInteger("amount") == 0 && !thisOre.getBoolean("lock")) {
                    ores.removeTag(oreName);
                }
                tag.setTag("ores", ores);
                itemStack.setTagCompound(tag);
            }
        }

        return EnumActionResult.PASS;
    }

    public EnumActionResult depositOresRandomly(IItemHandlerModifiable inventory, ItemStack itemStack) {
        NBTTagCompound tag = itemStack.getTagCompound();
        if (tag == null)
            return EnumActionResult.PASS;

        Random rand = new Random();
        List<StoredOre> ores = Lists.newArrayList();

        NBTTagCompound oresTag = tag.getCompoundTag("ores");
        for (String ore : oresTag.getKeySet()) {
            ores.add(this.getOre(itemStack, ore));
        }

        if (ores.isEmpty())
            return EnumActionResult.PASS;

        boolean done = false;
        while (!done) {
            StoredOre ore = ores.get(rand.nextInt(ores.size()));
            int amount = Math.min(ore.amount, rand.nextInt(45));

            ItemStack stack = DeepPocketsConfig.getStack(ore, amount);

            List<Integer> possibleSlots = Lists.newArrayList();
            for (int i = 0; i < inventory.getSlots(); i++) {
                ItemStack returned = inventory.insertItem(i, stack, true);
                if (returned != stack) {
                    if (inventory.getStackInSlot(i).isEmpty()) {
                        possibleSlots.add(i);
                    } else if (ItemHandlerHelper.canItemStacksStack(inventory.getStackInSlot(i), stack)) {
                        ItemStack slotStack = inventory.getStackInSlot(i);
                        if (slotStack.getCount() < slotStack.getMaxStackSize())
                            possibleSlots.add(i);
                    }
                }
            }

            if (possibleSlots.isEmpty()) {
                ores.remove(ore);
                this.setOre(itemStack, ore.oreDictName, ore);

                if (ores.isEmpty())
                    done = true;

                continue;
            }

            int index = possibleSlots.get(rand.nextInt(possibleSlots.size()));
            ItemStack slotStack = inventory.getStackInSlot(index);


            if (slotStack.isEmpty()) {
                inventory.setStackInSlot(index, stack.copy());
                ore.amount -= amount;
            } else if (ItemHandlerHelper.canItemStacksStack(slotStack, stack)) {
                int amountToTransfer = Math.min(amount, inventory.getSlotLimit(index) - slotStack.getCount());
                ItemStack toTransfer = stack.splitStack(amountToTransfer);

                int fullAmount = toTransfer.getCount() + slotStack.getCount();
                inventory.getStackInSlot(index).setCount(fullAmount); // TODO - Change this. It breaks contract
                ore.amount -= toTransfer.getCount();
            }

            if (ore.amount == 0) {
                ores.remove(ore);
                this.setOre(itemStack, ore.oreDictName, ore);

                if (ores.isEmpty())
                    done = true;
            }

        }


        return EnumActionResult.PASS;
    }

    public boolean containsOre(ItemStack pack, String oreName) {
        if (pack.getTagCompound() == null)
            return false;

        return pack.getTagCompound().getCompoundTag("ores").hasKey(oreName);
    }

    public int getOreAmount(ItemStack pack, String oreName) {
        if (pack.getTagCompound() == null || !containsOre(pack, oreName))
            return 0;

        return pack.getTagCompound().getCompoundTag("ores").getCompoundTag(oreName).getInteger("amount");
    }

    public int numStoredOres(ItemStack pack) {
        if (pack.getTagCompound() == null)
            return 0;

        return pack.getTagCompound().getCompoundTag("ores").getSize();
    }

    public StoredOre getOre(ItemStack pack, String oreName) {
        if (pack.getTagCompound() == null || !pack.getTagCompound().getCompoundTag("ores").hasKey(oreName)) {
            return new StoredOre(oreName, 0, "");
        }

        return new StoredOre(pack.getTagCompound().getCompoundTag("ores").getCompoundTag(oreName), oreName);
    }

    public void setOre(ItemStack pack, String oreName, StoredOre ore) {
        if (pack.getTagCompound() == null)
            pack.setTagCompound(new NBTTagCompound());

        NBTTagCompound oreTag = pack.getTagCompound().getCompoundTag("ores");
        ore.save(oreTag);
        pack.getTagCompound().setTag("ores", oreTag);
    }

    public boolean isEmpty(ItemStack pack) {
        NBTTagCompound tag = pack.getTagCompound();
        if (tag == null)
            return true;

        NBTTagCompound oresTag = tag.getCompoundTag("ores");
        for (String ore : oresTag.getKeySet()) {
            if (tag.getCompoundTag("ores").getCompoundTag(ore).getInteger("amount") > 0)
                return false;
        }

        return true;
    }

    public int getTotalItems(ItemStack pack) {
        NBTTagCompound tag = pack.getTagCompound();
        if (tag == null)
            return 0;

        int items = 0;
        NBTTagCompound oresTag = tag.getCompoundTag("ores");
        for (String ore : oresTag.getKeySet()) {
            items += tag.getCompoundTag("ores").getCompoundTag(ore).getInteger("amount");
        }

        return items;
    }

    public class StoredOre {
        public String oreDictName;
        public String registryName;
        public int meta;
        public int amount;
        public String displayName;
        public boolean locked;

        public StoredOre(String oreDictName, int amount, String displayName, boolean locked) {
            this.oreDictName = oreDictName;
            this.amount = amount;
            this.displayName = displayName;
            this.locked = locked;
        }

        public StoredOre(String oreDictName, int amount, String displayName) {
            this(oreDictName, amount, displayName, false);
        }

        public StoredOre(NBTTagCompound tag, String oreDictName) {
            this.oreDictName = oreDictName;
            this.registryName = tag.getString("registry");
            this.meta = tag.getInteger("meta");
            this.amount = tag.getInteger("amount");
            this.displayName = tag.getString("name");
            this.locked = tag.getBoolean("lock");
        }

        public void save(NBTTagCompound tag) {
            if (amount > 0 || locked) {
                NBTTagCompound oreTag = new NBTTagCompound();
                oreTag.setString("registry", registryName);
                oreTag.setInteger("meta", meta);
                oreTag.setInteger("amount", amount);
                oreTag.setString("name", displayName);
                oreTag.setBoolean("lock", locked);
                tag.setTag(oreDictName, oreTag);
            } else {
                tag.removeTag(oreDictName);
            }
        }
    }
}
